### ISSUE TITLE
question about training of entropybottleneck

### ISSUE THREAD

**Issue by wl3b10s** [InterDigitalInc/CompressAI#334]
I am trying to train a entropybottleneck for entropy coding with directly integer input.

there are no analysis and synthesis models before and after entropy coding in my model.

and my code for training are as follows:

    entropy = compressai.entropy_models.entropy_models.EntropyBottleneck
    my_entropy = entropy(1).double().to(device) # ch=1 input

    optimizer = torch.optim.Adam(my_entropy.parameters(), lr=1e-4, betas=(0.8, 0.99))
    aux_optimizer = torch.optim.Adam(my_entropy.parameters(), lr=1e-3, betas=(0.8, 0.99))    
    
    for epoch in range(0, 50000 + 1):
        print('epoch ', epoch)
        my_entropy.train()
        for data in train_loader: 
        
            in = data.to(device)

            optimizer.zero_grad()
            aux_optimizer.zero_grad()    
            
            outputs, likelihood = my_entropy(in.double())

            loss = torch.log(likelihood).sum() / (-math.log(2)) # loss only contains entropy estimation
            loss.backward()
            optimizer.step()
            
            aux_loss = my_entropy.loss()
            aux_loss.backward()
            aux_optimizer.step() 
            
            my_entropy.update() 

when i test the model for each epoch, it seems the result doesn't change. although lossless entropy coding is finished correctly. 

        string = my_entropy.compress(test.double())
        print(string[0].__sizeof__())  -> result string length doesn't change along training

so ,whether entropybottleneck can be trained in this way? or maybe there are some problem for my code to use entropybottleneck?

thanks.

**Comment by YodaEmbedding** [comment]
That looks mostly fine to me. I suggest two changes:

1. Use `.update(force=True)`.
2. Use `num_bits = 8 * len(string[0])`.

---

Full example:

```
import compressai
import torch
from compressai.entropy_models import EntropyBottleneck
from torch.utils.data import DataLoader, Dataset


class DummyDataset(Dataset):
    def __init__(self, *, num_samples, shape):
        self.num_samples = num_samples
        self.shape = shape

    def __len__(self):
        return self.num_samples

    def __getitem__(self, idx):
        return torch.randn(self.shape)


device = "cpu"

model = EntropyBottleneck(1)
model.to(device)

optimizer = torch.optim.Adam(model.parameters(), lr=1e-4)
aux_optimizer = torch.optim.Adam(model.parameters(), lr=1e-3)

train_dataset = DummyDataset(num_samples=640, shape=(1, 32))
test_dataset = DummyDataset(num_samples=64, shape=(1, 32))

train_loader = DataLoader(train_dataset, batch_size=16, shuffle=True)
test_loader = DataLoader(test_dataset, batch_size=64, shuffle=False)


for epoch in range(8):
    print(f"Epoch {epoch + 1}")

    model.train()

    for batch in train_loader:
        input = batch.to(device)

        optimizer.zero_grad()
        aux_optimizer.zero_grad()

        outputs, likelihood = model(input)

        loss = -likelihood.log2().sum()
        loss.backward()
        optimizer.step()

        aux_loss = model.loss()
        aux_loss.backward()
        aux_optimizer.step()

    model.eval()
    model.update(force=True)

    batch = next(iter(test_loader))
    input = batch.to(device)

    string = model.compress(batch)
    num_bits = 8 * sum(len(s) for s in string) / len(string)
    print(f"Bitrate: {num_bits:.2f}")
```

**Comment by wl3b10s** [comment]
thanks for your help.

it seems after adding '.update(force=True).', the bitrate during training decreased.
